File: python/Heap or Priority Queue/task_scheduler.py
from typing import List
from collections import Counter, deque
import heapq

class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        """
        Find the least number of units of times that the CPU will take to finish all the given tasks with cooldown n.
        Approach:
        - Use a max-heap to always schedule the most frequent task available.
        - Use a queue to keep track of tasks in cooldown (with their next available time).
        - At each time unit, either schedule a task or idle.
        - When a task's cooldown is over, push it back to the heap if it still has remaining count.
        Time Complexity: O(N), where N is the number of tasks
        Space Complexity: O(1) (since there are at most 26 tasks)
        """
        count = Counter(tasks)
        maxHeap = [-cnt for cnt in count.values()]
        heapq.heapify(maxHeap)

        time = 0
        q = deque() #pairs of [cnt, idleTime]
        
        while maxHeap or q:
            time += 1

            if maxHeap:
                cnt = 1 + heapq.heappop(maxHeap)
                if cnt:
                    q.append([cnt, time + n])

            if q and q[0][1] == time:
                heapq.heappush(maxHeap, q.popleft()[0])
        return time

# -------------------
# Test Cases
# -------------------
def test_task_scheduler():
    s = Solution()
    # Test case 1: Example
    tasks = ["A","A","A","B","B","B"]
    n = 2
    assert s.leastInterval(tasks, n) == 8  # A B idle A B idle A B
    # Test case 2: No cooldown
    tasks = ["A","A","A","B","B","B"]
    n = 0
    assert s.leastInterval(tasks, n) == 6
    # Test case 3: All unique tasks
    tasks = ["A","B","C","D"]
    n = 2
    assert s.leastInterval(tasks, n) == 4
    # Test case 4: Single task
    tasks = ["A","A","A","A"]
    n = 3
    assert s.leastInterval(tasks, n) == 13
    # Test case 5: Empty task list
    tasks = []
    n = 2
    assert s.leastInterval(tasks, n) == 0
    print("All test cases passed!")

if __name__ == "__main__":
    test_task_scheduler()

File: C++/Heap or Priority Queue/task_scheduler.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /**
     * Finds the least number of time units to finish all tasks with cooldown n.
     *
     * Approach:
     * - Use a max-heap to always select the task with the highest frequency.
     * - Use a queue to keep track of tasks in their cooldown period along with the time they become available again.
     * - At each time unit, either execute a task from the heap or stay idle.
     * - When a task's cooldown is over, push it back to the heap if it still has remaining count.
     *
     * Time Complexity: O(N), where N is the number of tasks.
     * Space Complexity: O(1), because there are at most 26 different tasks.
     */
    int leastInterval(vector<char>& tasks, int n) {
        unordered_map<char, int> count; 
        for (char task : tasks) {
            count[task]++; // count frequency of each task
        }

        // Max-heap to store task frequencies (negated to simulate max-heap in C++)
        priority_queue<int> maxHeap;
        for (auto& it : count) {
            maxHeap.push(it.second);
        }

        queue<pair<int, int>> q; // queue of {remaining count, available_time}

        int time = 0;

        while (!maxHeap.empty() || !q.empty()) {
            time++;

            if (!maxHeap.empty()) {
                int cnt = maxHeap.top() - 1; // execute the task, decrease its count
                maxHeap.pop();

                if (cnt > 0) {
                    q.push({cnt, time + n}); // put it in cooldown if it still needs to run
                }
            }

            if (!q.empty() && q.front().second == time) {
                maxHeap.push(q.front().first);
                q.pop();
            }
        }

        return time;
    }
};

/*
---------------------- LINE-BY-LINE SYNTAX EXPLANATION ----------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers (vector, queue, unordered_map, etc.)

using namespace std;
    ✅ So we don’t have to write std:: repeatedly.

class Solution { ... }
    ✅ Defines a class named Solution.

public:
    ✅ Makes the following methods accessible from outside the class.

int leastInterval(vector<char>& tasks, int n)
    ✅ Function that takes a vector of characters (task names) and an integer n.

unordered_map<char, int> count;
    ✅ Hash map to store the count of each task.

for (char task : tasks) { count[task]++; }
    ✅ Count the frequency of each task.

priority_queue<int> maxHeap;
    ✅ Max-heap that stores task counts.

for (auto& [task, cnt] : count) { maxHeap.push(cnt); }
    ✅ Push each task count into the maxHeap.

queue<pair<int, int>> q;
    ✅ Queue to store tasks in cooldown along with the time they become available.

int time = 0;
    ✅ Time counter.

while (!maxHeap.empty() || !q.empty()) { ... }
    ✅ Loop until both the heap and queue are empty.

time++;
    ✅ Increment time unit.

if (!maxHeap.empty()) { ... }
    ✅ If there is a task available, run it.

int cnt = maxHeap.top() - 1; maxHeap.pop();
    ✅ Run task and decrement count.

if (cnt > 0) q.push({cnt, time + n});
    ✅ If the task still has remaining executions, push it into cooldown queue.

if (!q.empty() && q.front().second == time) { ... }
    ✅ If a task's cooldown has finished, re-add it to the heap.

return time;
    ✅ Return the total time taken.
*/
